vumi/transports/dmark/dmark_ussd.py

```diff
@@ -191,12 +198,54 @@ def handle_outbound_message(self, message):
 
         response_id = self.finish_request(
             message['in_reply_to'], json.dumps(response_data))
+
+        yield self._set_request_end(message['in_reply_to'])
+
         if response_id is not None:
-            return self.publish_ack(
+            ack = yield self.publish_ack(
                 user_message_id=message['message_id'],
                 sent_message_id=message['message_id'])
+            returnValue(ack)
         else:
-            return self.publish_nack(
+            nack = yield self.publish_nack(
                 user_message_id=message['message_id'],
                 sent_message_id=message['message_id'],
                 reason="Could not find original request.")
+            returnValue(nack)
+
+    def _set_request_start(self, message_id):
+        self.session_timestamps[message_id] = time.time()
+    
+    def _set_request_end(self, message_id):
+        start_time = self.session_timestamps.pop(message_id, None)
+        if start_time:
+            response_time = time.time() - start_time
+            if response_time > 10:
+                return self.add_status(
+                    component='response',
+                    status='down',
+                    type='slow_response',
+                    message='Very slow response',
+                    reasons=['Response took longer than 10 seconds'],
+                    details={
+                        'response_time': response_time,
+                    })
+            elif response_time > 1:
+                return self.add_status(
+                    component='response',
+                    status='degraded',
+                    type='slow_response',
+                    message='Slow response',
+                    reasons=['Response took longer than 1 second'],
+                    details={
+                        'response_time': response_time,
+                    })
+            else:
+                return self.add_status(
+                    component='response',
+                    status='ok',
+                    type='response_sent',
+                    message='Response sent',
+                    details={
+                        'response_time': response_time,
+                    })
```

Comment:
> We could do that. We could also maybe have the set_request_start and set_request_end functions in the base transport. We could even have it all called automatically, and have each transport just configure the response times.

I have a vague preference for the `self.add_status` calls happening in the transport itself instead of in the base rpc class, seems like something we might want to control per-transport (maybe some transports have special cases, or something). Don't mind us trying the completely automatic approach though.